state/statemetrics: state metrics collector

axw opened the pull request:
Introduce a Prometheus metric collector for
high-level state metrics. This collector will
gather metrics such as how many models and
machines are managed by the controller; and
how many users are defined in or known to the
controller.

anastasiamac commented on cmd/jujud/agent/machine.go:
```diff
@@ -985,6 +986,17 @@ func (a *MachineAgent) startStateWorkers(st *state.State) (worker.Worker, error)
 				return newUpgradeMongoWorker(st, a.machineId, a.maybeStopMongo)
 			})
+			a.startWorkerAfterUpgrade(runner, "statemetrics", func() (worker.Worker, error) {
+				return worker.NewSimpleWorker(func(stop <-chan struct{}) error {
```
Would it not be neater if it was a method call? rather than defining new worker inline? 
It would be nice to test it in isolation as well.

axw replied:
I've split out the code to a separate function. The agent is kind of a mess, testing it in isolation is currently impractical

anastasiamac commented on cmd/jujud/agent/machine.go:
```diff
@@ -1036,6 +1048,7 @@ func (a *MachineAgent) startStateWorkers(st *state.State) (worker.Worker, error)
 			a.startWorkerAfterUpgrade(singularRunner, "txnpruner", func() (worker.Worker, error) {
 				return txnpruner.New(st, time.Hour*2, clock.WallClock), nil
 			})
+
```
I am not sure that this newline adds to readability :D

axw replied:
Reverted

axw commented:
> My only concern is border-line philosophical :)... Up until now, "metrics" had a different meaning in Juju. Maybe it's worthwhile to be blatantly clear that this are performance/observability related metrics?

I don't know how to convey that in a name succinctly and unambiguously. It's easy to change, so I'll keep it in mind and change it if I come up with something.

axw commented:
**QA**

- bootstrapped
- added some models
- added some machines
- added some users
(checked that all the changes showed up in the metrics)
- disabled mongo
(checked that the juju_state_* metrics didn't show up)

wallyworld commented on state/statemetrics/statemetrics_test.go:
```diff
@@ -0,0 +1,239 @@
+		controllerAccess: permission.NoAccess,
+	}, {
+		tag:              names.NewUserTag("cayley@cambridge"),
+		deleted:          true,
```
can we add another user with disabled = true to cover all possibilities

axw replied:
Done.

axw commented on cmd/jujud/agent/machine.go:
```diff
@@ -1629,3 +1633,15 @@ func metricAPI(st api.Connection) (metricsmanager.MetricsManagerClient, error) {
 	return deployer.NewSimpleContext(agentConfig, st)
 }
+
+func newStateMetricsWorker(st *state.State, registry *prometheus.Registry) worker.Worker {
```
Updated existing featuretest.

axw commented:
$$merge$$

axw commented:
$$merge$$

axw commented:
$$retry$$
